docs-ref-services/key-vault.md

```diff
@@ -142,18 +111,140 @@ operation = client.vaults.create_or_update(
+
+<pre>
+No credential in this chain provided a token.
+Attempted credentials:
+        EnvironmentCredential: Incomplete environment configuration
+        ImdsCredential: IMDS endpoint unavailable
+        SharedTokenCacheCredential: The shared cache contains no signed-in accounts.
+        To authenticate with SharedTokenCacheCredential, login through developer tooling
+        supporting Azure single sign on.
+</pre>
+
+There are two options to run the code locally:
+
+- Use a developer tool, like Visual Studio, that supports a single sign-on experience for Azure and therefore provides a SharedTokenCacheCredential.
+
+- Configure a local service principal and create environment variables to provide the EnvironmentCredential:
+
+    1. Install the [Azure CLI](/cli/azure/install-azure-cli?view=azure-cli-latest) if you don't have it installed already.
+
+    1. In a terminal or command prompt, create a service principle using the Azure CLI [az ad sp create-for-rbac](/cli/azure/ad/sp?view=azure-cli-latest#az-ad-sp-create-for-rbac) command:
+
+        ```azurecli
+        az ad sp create-for-rbac --skip-assignment
+        ```
 
-[1]: https://azure.microsoft.com/resources/samples/key-vault-python-manage/
-[2]: https://azure.microsoft.com/resources/samples/key-vault-recovery-python/
+        The command returns a JSON object similar to the following (your values will differ):
+
+        <pre>
+        {
+          "appId": "8a1f0000-97c0-4f78-a750-9cb0000c81f9",
+          "displayName": "azure-cli-2020-03-10-16-38-58",
+          "name": "http://azure-cli-2020-03-10-16-38-58",
+          "password": "05000052-3d77-4def-a457-53f428744af9",
+          "tenant": "58efb1eb-d777-46f6-9479-0000a50bf2f3"
+        }
+        </pre>
+
+    1. Create the following environment variables using values from the JSON:
+
+        | Variable name | Value from create-for-rbac output |
+        | --- | --- |
+        | `AZURE_CLIENT_ID` | `appId` |
+        | `AZURE_CLIENT_SECRET` | `password` |
+        | `AZURE_TENANT_ID` | `tenant` |
+
+    1. Grant the service principal access to the key vault, replacing `<your-resource-group>` with resource group containing your key vault, `<your-keyvault-name>` with the name of the Key Vault, and `<client-id>` with the `appID` value from the previous JSON output:
+
+        ```azurecli
+        az keyvault set-policy -g <your-resource-group> -n <your-keyvault-name> --spn <your-clientId> --secret-permissions delete get list set --key-permissions create decrypt delete encrypt get list unwrapKey wrapKey
+        ```
+
+        When using the management APIs, this command obviously requires that the key vault already exists. In that case, you must create the key vault using the [Azure CLI](/azure/key-vault/quick-create-cli), [Azure PowerShell](/azure/key-vault/quick-create-powershell), or the [Azure Portal](/azure/key-vault/quick-create-portal) instead. You can then perform other management tasks through code.
```

Comment:
This feels incongruous. The document began by walking me through creating a service principal with the CLI, then a Key Vault with the management library. Now it's walking me through creating another service principal and Key Vault, maybe with CLI, but possibly PowerShell or the Portal.